README

###############################################################################
 # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * #
###############################################################################

This document describes MCE version 1.600.

 Many-Core Engine (MCE) for Perl helps enable a new level of performance by
 maximizing all available cores. MCE spawns a pool of workers and therefore
 does not fork a new process per each element of data. Instead, MCE follows
 a bank queuing model. Imagine the line being the data and bank-tellers the
 parallel workers. MCE enhances that model by adding the ability to chunk
 the next n elements from the input stream to the next available worker.

 MCE uses domain sockets for IPC versus pipes. One day, a physical box will
 have hundreds of cores. MCE can spawn 960 workers (via fork) for ulimit re-
 porting 1024 allowed for max user processes. Other parallel implementations
 using pipes cannot do this (only about half of ulimit -u).

Chunking in MCE.

 Input data is optional in MCE. Thus, input data is not required to run MCE.

 MCE iterates over input data in chunks. Workers can be spawned prior to
 creating or obtaining data. MCE is a chunking engine. It does not divide
 the input equally by the number of workers. Instead, it chunks from
 start till end.

 Imagine input data as a highway. Now think of MCE (automobile) on the
 highway. Basically, data can be larger than physical memory allows.
 The chunking nature of MCE makes this possible.

    input_data => '/path/to/file',  ## process a file in parallel
    input_data => \@array,          ## process an array in parallel
    input_data => \*FileHandle,     ## process a filehandle in parallel
    input_data => \$scalar,         ## treated like a memory-based file
    input_data => \&iterator,       ## allows for custom input iterator

 MCE can iterate over a sequence of numbers mathematically if all you need
 is a numeric iterator. Check out forseq.pl and seq_demo.pl.

    sequence => { begin => 1, end => 100, step => 2 },
    chunk_size => 20,      ## Worker receives the next 20 sequences

Use Cases for MCE.

 1. Perl is not compiled with threads support and wanting something similar
    to Thread::Queue. Check out MCE::Flow and MCE::Queue.

 2. Wanting multiple roles and each role uniquely configured with number of
    workers and function to execute. Check out flow_demo.pl and step_demo.pl.

 3. Wanting to process rows from a database in parallel. Check out sampledb
    under the examples directory.

 4. MCE can be used to process a single (large) file or process many (small)
    files in parallel. Take bin/mce_grep, a wrapper around the grep binary,
    for a test drive. MCE also includes egrep.pl for a pure Perl grep
    implementation.

 5. A large telecom company wanting to poll millions of devices via SNMP.
    This is possible with MCE, AnyEvent::SNMP, and Net::SNMP. Think of
    having 200 workers and each worker connecting to 300 devices with
    AnyEvent. Each worker processes 300 devices simultaneously.

 6. Wanting to access many hosts in parallel via SSH or TCP/IP.

 ...

###############################################################################
 # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * #
###############################################################################

INSTALLATION

 To install this module type the following:

    MCE_INSTALL_TOOLS=1 perl Makefile.PL   (e.g. bin/mce_grep)

    (or) perl Makefile.PL

    make
    make test
    make install

DEPENDENCIES

 This module requires Perl 5.8.0 or later to run.

    MCE spawns child processes by default, not threads.

    However, MCE supports threads via 2 threading libraries if threads
    is desired.

    The use of threads in MCE requires that you include threads support
    prior to loading MCE. The use_threads option defaults to 1 when a
    thread library is loaded. Threads is loaded automatically for
    $^O eq 'MSWin32'.

    use threads;                use forks;
    use threads::shared;  (or)  use forks::shared;
    use MCE;                    use MCE;

 MCE utilizes the following modules:

    bytes
    constant
    Carp
    Fcntl
    File::Path
    IO::Handle
    Scalar::Util
    Socket
    Storable 2.04+
    Symbol
    Test::More 0.45+ (for testing only via make test)
    Time::HiRes

COPYRIGHT AND LICENCE

 Copyright (C) 2012-2015 by Mario E. Roy <marioeroy AT gmail DOT com>

 This program is free software; you can redistribute it and/or modify it
 under the terms of either: the GNU General Public License as published
 by the Free Software Foundation; or the Artistic License.

 See L<http://dev.perl.org/licenses/> for more information.

USAGE

 The source is hosted at http://code.google.com/p/many-core-engine-perl/

 https://metacpan.org/module/MCE::Signal
 https://metacpan.org/module/MCE

###############################################################################
 # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * # * #
###############################################################################

Enjoy MCE !!!

 - Mario



CHANGES

Revision history for MCE

1.699 Thu Mar 26 15:00:00 EST 2015

  [ TODO: before releasing MCE 1.700 ]

    Complete MCE::Shared documentation.
    Document ->do( qq/q( perl_text ) ).
    Document ->enq, ->enqp, and ->await methods in MCE::Step.
    Write MCE::Cookbook.pod.

  [ENHANCEMENTS]

  * Improved support for multi-level MCE instances; workers spawning MCE.
  * Optimized locking in Mutex and Queue for the (MSWin32) platform.

  * Locking for IPC is now handled by MCE::Mutex. This allows for more threads
    per process due to lesser consumption on file handles. This solves the slow
    file locking behavior seen in Cygwin.

  * Non-MCE processes or threads spawned from any MCE worker may share data
    between themselves and MCE. Queues from MCE::Queue are also supported
    along with MCE->do, MCE->gather, MCE->print, MCE->printf, MCE->say,
    and MCE->sendto. Examples are described in MCE::Cookbook.pod.

  * The MCE::Flow and MCE::Step Models can take an anonymous array for
    specifying use_threads uniquely for each sub-task. Mixing threads and
    processes is allowed across the board with this release. MCE on Cygwin
    has finally reached 100% parity with MCE on a UNIX OS.

  [NEW FEATURES]

  * MCE::Shared, MCE::Shared::Array, MCE::Shared::Hash, MCE::Shared::Scalar.
    Test scripts: 07_shared_[ array | hash | scalar ].t

  * Added examples/asynchrony demonstrating parallel concurrency, asynchrony,
    and shared data; dnslookup_ae.pl, dnslookup_ioa.pl, echoserver_ae.pl,
    echoserver_ioa.pl, echoserver_mojo.pl, and echoserver_poe.pl.

  * Added support for ->do( qq/q( perl_text ) ).
  * Replaced netping.pl with fastping_ae.pl supporting IPv4 and IPv6.
  * Added examples: shared_mce.pl, shared_thr.pl, and step_mon.pl.

  * Added ->enq, ->enqp, and ->await methods to MCE::Step.
  * Added ->await method to MCE::Queue.

1.604 Sat Mar 21 21:00:00 EST 2015

  This is the last release for the 1.6 branch.

  [BUG FIXES]

  * All bugs found during testing of the upcoming 1.7 release have been
    backported to the 1.6 branch.

  [NEW FEATURES]

  * Added out_iter_array and out_iter_fh to MCE::Candy. These preserve output
    order and cover the two general use cases.

1.603 Tue Mar 17 21:00:00 EST 2015

  [ENHANCEMENTS]

  * A safer solution by Dmitry Karasik for the die handler in MCE::Signal
    and MCE::Core::Worker.

  * Moved ->forchunk, ->foreach, and ->forseq sugar methods to MCE::Candy.
    Stubs exist in MCE. Thus, no breakage to existing apps.

  * Removed the link to MCE::Shared on the main page. I decided to backport
    all the fixes into 1.6. The MCE::Shared link was missed and requires
    the upcoming 1.7 release.

1.602 Mon Mar 16 20:00:00 EST 2015

  [BUG FIXES]

  * Updated die handler in MCE::Core::Worker and MCE::Signal (RT#102802)
    Bug reported by Dmitry Karasik.

  [ENHANCEMENTS]

  * MCE child processes call CORE::exit during exiting; previously CORE::kill.
  * Improved IPC stability on Windows and Cygwin.

  * Multiple calls to mutex->lock by the same process or thread is now safe.
    The mutex will remain locked until mutex->unlock is called.

1.601 Sun Mar 15 21:00:00 EST 2015

  This release is attributed to the many use cases sent by George Bouras.

  [BUG FIXES]

  * Updated Makefile.PL allowing installation of bin/mce_grep (RT#102040).
    Running (MCE_INSTALL_TOOLS=1 perl Makefile.PL) installs bin/mce_grep.

  * Ensure MCE instances spawned by workers have shutdown before leaving.
  * An exiting forked non-MCE process will not cause the worker to exit.

  * Fixed sockets not closing immediately in Cygwin and Windows.
  * Fixed a pod error in MCE::Mutex.

  * Fixed a rare condition where socket handles for MCE::Mutex and
    MCE::Queue were closing pre-maturely when using threads.

  * Fixed an undefined variable inside croak statements in MCE Models'
    import subroutine.

  * Fixed automatic shutdown due to loading threads and specifying
    use_threads => 0 with workers persisting before exiting the script.

  [ENHANCEMENTS]

  * Moved relay methods, introduced in 1.600, from MCE to MCE::Relay.
    The MCE init_relay option loads and enables MCE::Relay automatically.

  * Captured metrics from Linux (previously OS X) for forchunk.pl,
    foreach.pl, and forseq.pl. Updated Examples.pod.

  * Default to 'auto' for max_workers in bin/mce_grep, examples/egrep.pl,
    forchunk.pl, foreach.pl, forseq.pl, iterator.pl, and wc.pl.

  * Moved _create_socket_pair from MCE to MCE::Util as _make_socket_pair.
    Added _destroy_sockets. Updated MCE, Mutex, and Queue to use MCE::Util.

  * Added a CARP_NOT line to MCE Models.
  * Added support for ->next and ->last from any sub-tasks in MCE::Step.
  * Reverted a small change applied in 1.600 to ->do and ->gather.
  * The 'synchronize' method in Mutex is wantarray aware.
  * Updated POD header lines for method names; from =item to =head2.
  * Workers now set STDERR and STDOUT to flush automatically.

  [NEW FEATURES]

  * Added ->pid method to MCE.

1.600 Sat Jan 31 20:00:00 EST 2015

  [BUG FIXES]

  * Die handlers (in MCE::Signal and MCE::Core::Worker) are finally 100%.
    Furthermore, eval { die ... } statements behave correctly while running
    MCE itself inside an eval block. Thus, MCE on iPerl on top of iPython
    is possible via Devel::IPerl.

  * Fixed queues stalling from running (MCE::Queue fast => 1) on Linux.

  * MCE Models now set $MCE::FREEZE, $MCE::THAW, and $MCE::TMP_DIR when
    overriding freeze, thaw, and tmp_dir respectively at load time.

  * Pressing CTRL-D now ends STDIN the first time.

  [ENHANCEMENTS]

  * Added seven names to CREDITS; David Farrell, Demian Riccardi,
    Hisham Eldai, Joel Berger, Oleksandr Kharchenko, Wei Shen, and
    Zakariyya Mughal.

  * Refactored ->print, ->printf, and ->say. Optimized ->print some more.
  * Shorten $_queue to $_Q in Queue. ID is always sent first during IPC.
  * The init method in MCE Models can now take an array of options.


Makefile.PL

# Module makefile for MCE (using ExtUtils::MakeMaker)

use 5.008;

use strict;
use warnings;

use ExtUtils::MakeMaker;

my @exe_files;

if ($ENV{MCE_INSTALL_TOOLS}) {
   push @exe_files, 'bin/mce_grep';
}

WriteMakefile(
   ABSTRACT       => 'Many-Core Engine for Perl providing parallel processing capabilities',

   AUTHOR         => 'Mario E. Roy <marioeroy AT gmail DOT com>',

   NAME           => 'MCE',

   PREREQ_PM => {
      'bytes'        => 0,
      'constant'     => 0,
      'Carp'         => 0,
      'Fcntl'        => 0,
      'File::Path'   => 0,
      'Getopt::Long' => 0,
      'IO::Handle'   => 0,
      'Scalar::Util' => 0,
      'Socket'       => 0,
      'Storable'     => 2.04,
      'Symbol'       => 0,
      'Time::HiRes'  => 0
   },

   VERSION        => '1.699',

   EXE_FILES      => [ @exe_files ],

   (($ExtUtils::MakeMaker::VERSION lt '6.25') ?
      (PL_FILES => { }) : ()),

   (($ExtUtils::MakeMaker::VERSION ge '6.31') ?
      (LICENSE => 'perl') : ()),

   (($ExtUtils::MakeMaker::VERSION ge '6.46') ?
      (META_MERGE => {
         build_requires => {
            'ExtUtils::MakeMaker' => 0,
            'Test::More'          => 0.45
         },
         no_index => {
            'directory' => [
               'examples',
               'images'
            ],
            'file' => [
               'bin/mce_grep'
            ]
         },
         resources => {
            'homepage'   => 'http://code.google.com/p/many-core-engine-perl/',
            'repository' => 'http://code.google.com/p/many-core-engine-perl/',
            'license'    => 'http://dev.perl.org/licenses/'
         },
         provides => {
            'MCE' => {
               'file'    => 'lib/MCE.pm',
               'version' => '1.699'
            },
            'MCE::Candy' => {
               'file'    => 'lib/MCE/Candy.pm',
               'version' => '1.699'
            },
            'MCE::Core::Input::Generator' => {
               'file'    => 'lib/MCE/Core/Input/Generator.pm',
               'version' => '1.699'
            },
            'MCE::Core::Input::Handle' => {
               'file'    => 'lib/MCE/Core/Input/Handle.pm',
               'version' => '1.699'
            },
            'MCE::Core::Input::Iterator' => {
               'file'    => 'lib/MCE/Core/Input/Iterator.pm',
               'version' => '1.699'
            },
            'MCE::Core::Input::Request' => {
               'file'    => 'lib/MCE/Core/Input/Request.pm',
               'version' => '1.699'
            },
            'MCE::Core::Input::Sequence' => {
               'file'    => 'lib/MCE/Core/Input/Sequence.pm',
               'version' => '1.699'
            },
            'MCE::Core::Manager' => {
               'file'    => 'lib/MCE/Core/Manager.pm',
               'version' => '1.699'
            },
            'MCE::Core::Validation' => {
               'file'    => 'lib/MCE/Core/Validation.pm',
               'version' => '1.699'
            },
            'MCE::Core::Worker' => {
               'file'    => 'lib/MCE/Core/Worker.pm',
               'version' => '1.699'
            },
            'MCE::Flow' => {
               'file'    => 'lib/MCE/Flow.pm',
               'version' => '1.699'
            },
            'MCE::Grep' => {
               'file'    => 'lib/MCE/Grep.pm',
               'version' => '1.699'
            },
            'MCE::Loop' => {
               'file'    => 'lib/MCE/Loop.pm',
               'version' => '1.699'
            },
            'MCE::Map' => {
               'file'    => 'lib/MCE/Map.pm',
               'version' => '1.699'
            },
            'MCE::Mutex' => {
               'file'    => 'lib/MCE/Mutex.pm',
               'version' => '1.699'
            },
            'MCE::Queue' => {
               'file'    => 'lib/MCE/Queue.pm',
               'version' => '1.699'
            },
            'MCE::Relay' => {
               'file'    => 'lib/MCE/Relay.pm',
               'version' => '1.699'
            },
            'MCE::Shared' => {
               'file'    => 'lib/MCE/Shared.pm',
               'version' => '1.699'
            },
            'MCE::Shared::Array' => {
               'file'    => 'lib/MCE/Shared/Array.pm',
               'version' => '1.699'
            },
            'MCE::Shared::Hash' => {
               'file'    => 'lib/MCE/Shared/Hash.pm',
               'version' => '1.699'
            },
            'MCE::Shared::Scalar' => {
               'file'    => 'lib/MCE/Shared/Scalar.pm',
               'version' => '1.699'
            },
            'MCE::Signal' => {
               'file'    => 'lib/MCE/Signal.pm',
               'version' => '1.699'
            },
            'MCE::Step' => {
               'file'    => 'lib/MCE/Step.pm',
               'version' => '1.699'
            },
            'MCE::Stream' => {
               'file'    => 'lib/MCE/Stream.pm',
               'version' => '1.699'
            },
            'MCE::Subs' => {
               'file'    => 'lib/MCE/Subs.pm',
               'version' => '1.699'
            },
            'MCE::Util' => {
               'file'    => 'lib/MCE/Util.pm',
               'version' => '1.699'
            }
         }
      }) : ()),

   (($ExtUtils::MakeMaker::VERSION ge '6.48') ?
      (MIN_PERL_VERSION => 5.008) : ()),

   INSTALLDIRS => (($] < 5.011) ? 'perl' : 'site')
);

